### PR TITLE
SAK-47117 Gradebook import fails if 2 or more users have the same name with different IDs

### PR DESCRIPTION
**Pull request by bjones86**
https://sakaiproject.atlassian.net/browse/SAK-47117

JDK seems to be calling `compareTo` now for the sorted set. I tried implemented equals/hash on the existing `GbUser` object, but due to complexities with how we use this object, it became difficult to keep equals and hashcode in sync.

Instead, I decided to split the usage of `GbUser` into a new class, so we use `GbUser` when we know we have UUID etc, and we use the new `GbUnidentifiedUser` for when we don't have a UUID. This makes the equals/hash on each consistent and easy to implement.

Really we are only using the new object to report on users contained within the file who are not members of the site, so it was a small refactor.

Also added fallback logic to the EID in the compareTo.

### CODE REVIEW

**Review by ottenhoff — APPROVED** [review]
Looks okay. I will add a test for the UserSortName in a separate PR (forthcoming)

**Comment by bjones86** [comment]
Hmm I may have just found a bug.

**Comment by bjones86** [comment]
Seems the build didn't like my last changes to the Comparator. I'm open to suggestions.

File: kernel/kernel-util/src/main/java/org/sakaiproject/util/comparator/UserSortNameComparator.java [hunk 1]
```diff
@@ -56,6 +56,6 @@ public int compare(User u1, User u2) {
         if (u1 == null) return (nullsLow ? -1 : 1);
         if (u2 == null) return (nullsLow ? 1 : -1);
 
-        return new NullSafeComparator<>(collator, nullsLow).compare(u1.getSortName(), u2.getSortName());
+        return new NullSafeComparator<>(collator, nullsLow).compare(u1.getSortName() + u1.getDisplayId(), u2.getSortName() + u2.getDisplayId());
```
Comment (ottenhoff) [inline kernel/kernel-util/src/main/java/org/sakaiproject/util/comparator/UserSortNameComparator.java:59]: why dont you just drop this part from the PR and we can do it in my PR

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbUnidentifiedUser.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * This is a minimal class representing an unidentified user, containing only the display ID (EID) and display name.
+ * @author bjones86
+ */
+@Getter
+@EqualsAndHashCode
+public class GbUnidentifiedUser implements Serializable, Comparable<GbUnidentifiedUser>
```
Comment (ern) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbUnidentifiedUser.java:15]: Would suggest moving this DisplayId comparator in with the others, someone else may find it useful.

**Comment by ern** [comment]
Does simply filtering the nulls mae sense?

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbUnidentifiedUser.java (hunk 2 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/model/GbUnidentifiedUser.java:15]: This object has very limited usefulness in GB. If there was prolific use of this object I would agree, but it exists solely for a single purpose.

**Review by ottenhoff — APPROVED** [review]
I'm okay with us although I don't fully understand it. What's the answer to Earle's question re "why not just filter nulls"?

**Comment by bjones86** [comment]
Filtering the nulls does not allow us to report which users were found in the file who are not members of the site. Unless I'm confused as to where "filtering the nulls" applies to.

**Comment by bjones86** [comment]
I'm happy to explain anything that remains unclear.

**Comment by ottenhoff** [comment]
ok i understand now

ok looks good

**Pull request merged**